Source/EventFlow.Tests/UnitTests/EventStores/ConcurrentFilesEventPersistanceTests.cs

```diff
@@ -0,0 +1,211 @@
+            definitionService.Load(typeof(ThingyPingEvent));
+
+            _serializer = new EventJsonSerializer(new JsonSerializer(), definitionService, factory);
+        }
+
+        [SetUp]
+        public void CreateStoreRootDir()
+        {
+            _storeRootPath = Path.Combine(
+                Path.GetTempPath(),
+                Guid.NewGuid().ToString());
+
+            Directory.CreateDirectory(_storeRootPath);
+        }
+
+        [TearDown]
+        public void DeleteStoreRootDir()
+        {
+            Directory.Delete(_storeRootPath, true);
+        }
+
+        [Test]
+        public void MultipleInstancesWithSamePathFail()
+        {
+            // Arrange
+            var tasks = RunInParallel(async i =>
+            {
+                var persistence = CreatePersistence("SameForAll");
+                await CommitEvents(persistence).ConfigureAwait(false);
+            });
+
+            // Act
+            Action action = () => Task.WaitAll(tasks.ToArray());
+
+            // Assert
+            action.ShouldThrow<IOException>("because of concurrent access to the same files.");
+        }
+
+        [Test]
+        public void MultipleInstancesWithDifferentPathsWork()
+        {
+            // Arrange
+            var tasks = RunInParallel(async i =>
+            {
+                var persistence = CreatePersistence(i.ToString());
+                await CommitEvents(persistence).ConfigureAwait(false);
+            });
+
+            // Act
+            Action action = () => Task.WaitAll(tasks.ToArray());
+
+            // Assert
+            action.ShouldNotThrow();
+        }
+
+        [Test]
+        public void SingleInstanceWorks()
+        {
+            // Arrange
+            var persistence = CreatePersistence();
+            var tasks = RunInParallel(async i =>
+            {
+                await CommitEvents(persistence).ConfigureAwait(false);
+            });
+
+            // Act
+            Action action = () => Task.WaitAll(tasks.ToArray());
+
+            // Assert
+            action.ShouldNotThrow();
+        }
+
+        private IFilesEventStoreConfiguration ConfigurePath(string storePath)
+        {
+            var fullPath = Path.Combine(_storeRootPath, storePath);
+            return FilesEventStoreConfiguration.Create(fullPath);
+        }
+
+        private FilesEventPersistence CreatePersistence(string storePath = "")
+        {
+            var log = new NullLog();
+            var serializer = new JsonSerializer();
+            var config = ConfigurePath(storePath);
+            var locator = new FilesEventLocator(config);
+            return new FilesEventPersistence(log, serializer, config, locator);
+        }
+
+        private async Task CommitEvents(FilesEventPersistence persistence)
+        {
+            var events = Enumerable.Range(0, NumberOfEventsPerBatch)
+                .Select(i => new ThingyPingEvent(PingId.New))
+                .ToArray();
```

Comment:
Use `ToList()`

Reply:
TIL. I was using ToArray all the time as it intuitively seemed more lightweight. Now I will switch from array return types to IEnumerable and use ToList. Will take some time getting used to it.

Reply:
Same for me 😄

Reply:
In EventFlow i use `IEnumerable` for returns that are lazy evaluated and `IReadOnlyCollection` for stuff that's collections and already evaluated